--- a/oviewer/document.go
+++ b/oviewer/document.go
@@ -133,8 +133,6 @@ type Document struct {
        // bottomLX is the leftmost X position on the last line.
        bottomLX int

-       // startX is the start position of x.
-       startX int
        // startY is the start position of y.
        startY int
        // scrollX is the starting position of the current scrollX.
@@ -143,9 +141,15 @@ type Document struct {
        columnCursor int
        // columnStart is the starting position of the column.
        columnStart int
-       // bodyStartX is the start position of the document body, excluding the line number area.
+       // leftMargin is the left margin for body drawing.
+       leftMargin int
+       // rightMargin is the right margin for body drawing.
+       rightMargin int
+       // lineNumberWidth is the width of the line number area (0 is not displayed).
+       lineNumberWidth int
+       // bodyStartX is the actual start position of the body (leftMargin + lineNumberWidth)
        bodyStartX int
-       // bodyWidth is the width of the document body. excluding line number area.
+       // bodyWidth is the width of the document body (excluding left/right margin and line number area).
        bodyWidth int

        // lastSearchLN is the last search line number.
